ext/opcache/Optimizer/zend_inference.c

```diff
@@ -3443,18 +3414,10 @@ static int zend_update_type_info(const zend_op_array *op_array,
 						case ZEND_FETCH_DIM_RW:
 						case ZEND_FETCH_DIM_FUNC_ARG:
 						case ZEND_FETCH_LIST_W:
-						case ZEND_ASSIGN_ADD:
-						case ZEND_ASSIGN_SUB:
-						case ZEND_ASSIGN_MUL:
-						case ZEND_ASSIGN_DIV:
-						case ZEND_ASSIGN_MOD:
-						case ZEND_ASSIGN_SL:
-						case ZEND_ASSIGN_SR:
-						case ZEND_ASSIGN_CONCAT:
-						case ZEND_ASSIGN_BW_OR:
-						case ZEND_ASSIGN_BW_AND:
-						case ZEND_ASSIGN_BW_XOR:
-						case ZEND_ASSIGN_POW:
+						case ZEND_ASSIGN_OP:
+						case ZEND_ASSIGN_DIM_OP:
+						case ZEND_ASSIGN_OBJ_OP:
+						case ZEND_ASSIGN_STATIC_PROP_OP:
```

Comment:
This doesn't look right. I think at least the `ZEND_ASSIGN_OBJ_OP` case should be in the block below. `ZEND_ASSIGN_OP` and `ZEND_ASSIGN_STATIC_PROP_OP` probably can't occur.

It would be good to replace the `default` below with `EMPTY_SWITCH_DEFAULT_CASE()` to make sure all possibilities are covered, but we'd also have to skip this switch for FETCH_DIM_R/IS.

Reply:
Just checked that for

```
function test() {
    $ary = [];
    $ary[0]->y += 2;
    var_dump($ary);
}
```

we are indeed producing incorrect inference right now:

```
            #3.V1 [null] = FETCH_DIM_RW #1.CV0($ary) [array] -> #2.CV0($ary) [array [long] of [array]] int(0)
            ASSIGN_ADD (obj) #3.V1 [null] string("y")
```

It should be array of object.

Reply:
I didn't fix this as a part of this PR, because this should be fixed in early versions as well.
@nikic could you please take care about this.